### DISCUSSION TITLE
const() function not work in certain situation

### DISCUSSION THREAD

**Discussion by IlysvlVEizbr** [micropython/micropython#12207] [General]
When using the const function in the REPL environment according to the examples in the [official documentation](http://docs.micropython.org/en/latest/library/micropython.html#micropython.const), I encountered the following error:
```python
>>> from micropython import const
>>>
>>> CONST_X = const(123)
>>> CONST_Y = const(2 * CONST_X + 1)
Traceback (most recent call last):
  File "<stdin>", line 1
SyntaxError: not a constant
```
Other known scenarios that can trigger the same error include:
```python
CONST_Y = const(CONST_X + 1)
CONST_Y = const(CONST_X)
CONST_Y = const(int(4))
CONST_Y = const(1.23 * 10)
CONST_Y = const(4 / 2)
```
But the parameters of these const() are indeed constants (after calculated). At the same time, in the following examples, const() will not throw an error.
```python
CONST_Y = const(123 + 1)
CONST_Y = const(123)
CONST_Y = const(4)
CONST_Y = const(123)
CONST_Y = const(2)
```

Hardware I am using: Pyb v11
Firmware I am using: MicroPython v1.20.0 on 2023-04-26

**Comment by robert-hh** [comment]
Try the most recent nightly build, and you do not have to import const.

**Reply by IlysvlVEizbr** [reply]
I have confirmed that the issue with `const()` persists in the latest version v1.20.0-327-gd14ddcbdb.

**Comment by peterhinch** [comment]
This is not a MicroPython bug. `const` is a compiler directive and does not work at the REPL. The following works if you save it to a file and import it:
```python
CONST_X = const(123)
CONST_Y = const(CONST_X + 1)
print(CONST_Y)
```
`const` causes the compiler to emit the numeric value of the constant rather than a reference to a global variable. This is faster and more efficient.

**Reply by IlysvlVEizbr** [reply]
In that way, we will be back to the mode of transforming code into some file and writing it into the MCU, as well as overall debugging, which will greatly reduce the convenience of MicroPython.
I think that the REPL is a good place to debug code line by line, and we may need some compatibility measures to ensure that the `const` can maintain its original function in the REPL (even without performance optimization), so that various codes containing `const` can run normally in the REPL (as I said at the beginning, some `const` functions can already run well in REPL now).

**Reply by peterhinch** [reply]
It is possible to debug code without writing it to the MCU using [mpremote](http://docs.micropython.org/en/latest/reference/mpremote.html).

The nature of the REPL is that a line of code is compiled and run, returning you to a prompt. The compiler does not retain state between prompts (if it did, life would be very confusing). Consequently compiler directives are not retained. The only thing I can suggest is to define a do-nothing function
```python
myconst = lambda *_ : None
```
Use `myconst` in place of `const` while debugging. When you deploy your file do a global search and replace.

**Reply by jimmo** [reply]
@IlysvlVEizbr Unfortunately this is not possible to fix due to the way const is implemented (exactly as Peter says, the const table is internal compiler state). Peter's workaround is a good suggestion!
